org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/model/TypeReference.java

```diff
@@ -0,0 +1,296 @@
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.emf.ecore.EPackage;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+import org.eclipse.xtext.xtext.generator.IXtextGeneratorLanguage;
+import org.eclipse.xtext.xtext.generator.util.GenModelUtil2;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Iterables;
+
+/**
+ * Reference to a Java type. Use this for automatic import of types in
+ * {@link JavaFileAccess} and {@link XtendFileAccess}.
+ */
+public class TypeReference {
+	public static class QualifiedClassName {
+		private final String packageName;
+
+		private final String className;
+
+		public String getPackageName() {
+			return packageName;
+		}
+
+		public String getClassName() {
+			return className;
+		}
+
+		public QualifiedClassName(String packageName, String className) {
+			this.packageName = packageName;
+			this.className = className;
+		}
+	}
+
+	public static TypeReference typeRef(String name, TypeReference... arguments) {
+		return new TypeReference(name, arguments == null ? null : Arrays.asList(arguments));
+	}
+
+	/**
+	 * @deprecated this method is available for backwards compatibility reasons
+	 */
+	@Deprecated
+	public static TypeReference guessTypeRef(String name, TypeReference... arguments) {
+		return new TypeReference(name, arguments == null ? null : Arrays.asList(arguments), false);
+	}
+
+	public static TypeReference typeRef(Class<?> clazz, TypeReference... arguments) {
+		return new TypeReference(clazz, arguments == null ? null : Arrays.asList(arguments));
+	}
+
+	public static TypeReference typeRef(EClass clazz, IXtextGeneratorLanguage language) {
+		ResourceSet _resourceSet = language.getResourceSet();
+		return new TypeReference(clazz, _resourceSet);
+	}
+
+	private final String packageName;
+
+	private final List<String> simpleNames;
+
+	private final List<TypeReference> typeArguments;
+
+	public TypeReference(String qualifiedName) {
+		this(qualifiedName, ((List<TypeReference>) null));
+	}
+
+	public TypeReference(String qualifiedName, List<TypeReference> arguments) {
+		this(qualifiedName, arguments, true);
+	}
+
+	private TypeReference(String qualifiedName, List<TypeReference> arguments, boolean strict) {
+		this(TypeReference.getPackageName(qualifiedName, strict), TypeReference.getClassName(qualifiedName, strict),
+				arguments);
+	}
+
+	public TypeReference(String packageName, String className) {
+		this(packageName, className, null);
+	}
+
+	public TypeReference(String packageName, String className, List<TypeReference> arguments) {
+		if (packageName == null) {
+			throw new IllegalArgumentException("Invalid package name: " + packageName);
+		}
+		if (className == null) {
+			throw new IllegalArgumentException("Invalid class name: " + className);
+		}
+		this.packageName = packageName;
+		this.simpleNames = Arrays.asList(className.split("\\."));
```

Comment:
`this.` not needed here, and on the lines 109, 126, 127, 131, and 134.

Reply:
i kept it for consistency with place right next to it where it is needed cause of name clash with parameter